Fixes markus20213

markusvoelter opened the pull request:
A couple of smaller bugfixes and "quirk removals" in data tables and test vectors.
No really breaking changes. See discussion in Slack.

alexanderpann reviewed (changes requested):
Looks good, just one small change:

- The short description is missing for `CurrentDateExpr` and it should use the iets3Keyword

alexanderpann commented on code/languages/org.iets3.opensource/.mps/libraries.xml:
```diff
@@ -1,13 +1,13 @@
-              <option name='name' value='mbeddr.plattform' />
-              <option name='path' value='$PROJECT_DIR$/../../../build/dependencies/com.mbeddr.platform' />
+              <option name="name" value="mbeddr.plattform" />
+              <option name="path" value="${iets3.github.opensource.home}/build/dependencies/com.mbeddr.platform" />
```
I am fine with the change but let's not switch between the project dir variable and the path variable in other PRs anymore.

markusvoelter replied:
I have changed it back.

alexanderpann commented on code/languages/org.iets3.opensource/_spreferences/TestExecutionPreferences/models/org.iets3.opensource.__spreferences.TestExecutionPreferences.mps:
```diff
@@ -15,12 +15,12 @@
         <child id="4473287864570320758" name="executionMode" index="3ZOXzE" />
       </concept>
-      <concept id="4473287864570320840" name="org.iets3.core.expr.testExecution.structure.InterpreterExecutionMode" flags="ng" index="3ZOXxk" />
+      <concept id="4473287864570320810" name="org.iets3.core.expr.testExecution.structure.GeneratorExecutionMode" flags="ng" index="3ZOXwQ" />
```
We always had the execution mode set to the interpreter but again this change seems better for testing our languages anyway :+1:

markusvoelter replied:
Ok, then no change here. I will keep the generator.

alexanderpann reviewed (changes requested):
Thanks for the efforts, especially with writing a generator for the data table language:

*   PriorisationHelper: Please select a color for the Darcula theme with the constructor JBColor(regular,dark). We have another PR for Darcula support that we haven't merged yet.
*   Do you mind including the proposed fix in #691 for the lookupBy method as well as checking if it is compatible with your generator? The way the table lookup works right now is definitely wrong.
*   Adding the `genjava.advanced` devkit to the `datatable@tests` model, I see that the `EmptyValue` is not supported by the generator as well as the `equals expression` and the `not equals expression` in the test case tests.
*   e.get(col) in `reduce_DataTable` is not null safe and throws a NullPointerException if wrong values are supplied as arguments. Example for TestsForDataTableGenerator:
    *   assert Table.lookUpBy\<val1>(30)!.val2 equals 4

markusvoelter commented:
Regarding the more detailed comment:

- PriorisationHelper: I have changed it to a JBColor. I use the same color for both cases (see image at the bottom). However the color for the regular menu entries doesn't work. But I suspect that is the PR for Darcula that you were talking about.
- EmptyValue: I have added the generator for EmptyValue. Why were the tests green even though this was missing? I don't understand the equals/not equals expression thing though. Shouldn't this work generically if/when the expression is reduced to null? Also, same Q: why did this work until now?
- reduce_DataTable is null-safe.

Let's see if the build runs.

![image](https://github.com/IETS3/iets3.opensource/assets/592330/b1d71082-b477-4a08-8736-c90cac0d99cc)

alexanderpann commented:
You have to restart MPS when you change the theme because it isn't applied everywhere automatically. The color doesn't look nice when using the Darcula theme:

<img width="564" alt="Screenshot 2024-01-09 at 13 09 54" src="https://github.com/IETS3/iets3.opensource/assets/88385944/3040e306-fa38-4ef2-b63f-2e5c3320749a">


I just noticed that in `datatable@tests` the generated code for keys like `keyB` is also generated wrong when cells are null or empty values are used despite the constructor wanting two integer arguments in the example:
`public static final TableWithEmpties_Table_Entry keyB = new TableWithEmpties_Table_Entry(null);`
Exception:
```
constructor TableWithEmpties_Table_Entry in class test.in.expr.os.datatable.Generated_DataTable_Test.TableWithEmpties_Table_Entry cannot be applied to given types;
  required: java.lang.Number,java.lang.Number
  found: <nulltype>
  reason: actual and formal argument lists differ in length (/Applications/mps/platforms/2021.3/iets3.opensource_5/code/languages/org.iets3.opensource/tests/test.in.expr.os/test_gen/test/in/expr/os/datatable/Generated_DataTable_Test.java:181)
```

The reason why we didn't notice all those issues is that the devkit `genjava.advanced` is missing for `datatable@tests`, so the tests where only executed with the interpreter and not the generator on the CI. The issue with the equals expressions is that the type of the left and right side are not supported. This has nothing to do with the null values. Those reductions rules are not generic and are always specific for a type e.g: http://127.0.0.1:63320/node?ref=r%3A3490917a-a35f-43d7-9a1f-1311fda2da92%28main%40generator%29%2F8286534136181769422

markusvoelter commented:
> You have to restart MPS when you change the theme because it isn't applied everywhere automatically. 

Ok. Fixed. It's dark blue-ish now. 

> I just noticed that in datatable@tests the generated code for keys like 
> keyB is also generated wrong when cells are null or empty values are used 
> despite the constructor wanting two integer arguments in the example:

Fixed.

> The reason why we didn't notice all those issues is that the devkit genjava.advanced 
> is missing for datatable@tests, so the tests where only executed with the 
> interpreter and not the generator on the CI. 

Interesting that this doesn't fail somehow then. Fixed,

> The issue with the equals expressions is that the type of the left and 
> right side are not supported. 

Fixed.

This last test here
   WithDefault.lookUpBy<val1>(30)!.val2
cannot work. The lookup by 30 doesn't find anything, so lookup returns null
as it should. But we're forcing the null check away, so this has to result
in an NPE. Deleting the test.

alexanderpann reviewed (approved):
LGTM.
